packages/module/src/DetailsPageHeader/utils/ActionMenu.tsx

```diff
@@ -0,0 +1,202 @@
+  displayLabelBeforeIcon?: boolean;
+  /** Optional id for action menu (defaults to 'actions') */
+  id?: string;
+};
+
+export type ActionMenuProps = EitherNotBoth<
+  { actions: ActionProps[] },
+  { groupedActions: GroupedActionsProps[] }
+> &
+  ActionMenuOptions;
+
+const useStyles = createUseStyles({
+  menuItemWithLabelBeforeIcon: {
+    '&.pf-c-dropdown__menu-item': {
+      flexDirection: 'row-reverse',
+      justifyContent: 'left'
+    },
+    '.pf-c-dropdown__menu-item-icon': {
+      marginLeft: 'var(--pf-c-dropdown__menu-item-icon--MarginRight)'
+    },
+    '.pf-c-dropdown__menu-item-main': {
+      flexDirection: 'row-reverse',
+    }
+  }
+});
+
+export const ActionMenu: React.FunctionComponent<ActionMenuProps> = ({
+  actions = [],
+  groupedActions = [],
+  isDisabled,
+  variant = ActionMenuVariant.DROPDOWN,
+  label = 'Actions',
+  position = DropdownPosition.right,
+  displayLabelBeforeIcon,
+  id = 'actions',
+}) => {
+  const [ isOpen, setIsOpen ] = React.useState(false);
+  const [ isGrouped, setIsGrouped ] = React.useState(false);
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const [ dropdownActionItems, setDropdownActionItems ] = React.useState<any[]>([]);
+
+  const classes = useStyles();
+
+  React.useEffect(() => {
+    if (groupedActions.length > 0) {
+      setIsGrouped(true);
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  const onToggle = (open: boolean) => {
+    setIsOpen(open);
+  };
+
+  const onFocus = () => {
+    const element = document.getElementById(`toggle-menu-${id}`);
+    if (element) {
+      element.focus();
+    }
+  };
+
+  const onSelect = () => {
+    setIsOpen(false);
+    onFocus();
+  };
+
+  /** Returns a DropDownItem element corresponding to an action */
+  const dropdownActionItem = React.useCallback(
+    (action: ActionProps) => {
+      const externalIcon =
+        'href' in action.cta &&
+        'external' in action.cta &&
+        action.cta.href &&
+        action.cta.external ? (
+            <ExternalLinkAltIcon />
+          ) : null;
+      const icon = action.icon ?? externalIcon;
+      const href = 'href' in action.cta ? action.cta.href : undefined;
+      const onClick =
+        'callback' in action.cta && action.cta.callback ? action.cta.callback : undefined;
+      return (
+        <DropdownItem
+          className={displayLabelBeforeIcon ? classes.menuItemWithLabelBeforeIcon : ''}
+          key={action.id}
+          tooltip={action.tooltip}
+          description={action.description}
+          icon={icon}
+          href={href}
+          isDisabled={action.isDisabled}
+          onClick={onClick}
+        >
+          {action.label}
+        </DropdownItem>
+      );
+    },
+    [ classes.menuItemWithLabelBeforeIcon, displayLabelBeforeIcon ],
+  );
+
+  React.useEffect(() => {
```

Comment:
This does not have to effect/state combo. You can do this computation in `useMemo` hook and have them available even on the initial render.

Reply:
Updated.